reactor-extra/src/main/java/reactor/cache/CacheHelper.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2011-2017 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.cache;
+
+import java.util.Map;
+
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Signal;
+import reactor.util.annotation.NonNull;
+
+/**
+ * Cache helper that may be used with any cache vendors that has a Map wrapper support.
+ * <p>
+ * Example usage:
+ * <pre><code>
+ *    LoadingCache<Integer, Object> graphs = Caffeine.newBuilder()
+ *                                       .maximumSize(10_000)
+ *                                       .expireAfterWrite(5, TimeUnit.MINUTES)
+ *                                       .refreshAfterWrite(1, TimeUnit.MINUTES)
+ *                                       .build(key -> createExpensiveGraph(key));
+ *
+ *    keyStream.concatMap(key -> Cache.hydrateFrom(graphs.asMap(), key)
+ *                                   .or(repository.findOneById(key))
+ * </code></pre>
+ *
+ * @author Oleh Dokuka
+ */
+
+public class CacheHelper {
+
+	/**
+	 * Restore a {@link Mono<Value>} from the cache-map given a provided key. If no value
+	 * is in the cache, it will be calculated from the original source which is set up in
+	 * the next step. Note that if the source completes empty, this result will be cached
+	 * and all subsequent requests with the same key will return {@link Mono#empty()). The
+	 * behaviour is similar for erroring sources, except cache hits would then return
+	 * {@link Mono#error(Throwable)}.
+	 *
+	 * @param cache {@link Map} wrapper of a cache
+	 * @param key mapped key
+	 * @param <Key> Key Type
+	 * @param <Value> Value Type
+	 *
+	 * @return Lazy "Or" Builder
+	 */
+	@NonNull
+	public static <Key, Value> Or<Value> hydrateFrom(@NonNull Map<Key, Signal<? extends Value>> cache,
```

Comment:
Map's generic signature could be `Map<? super Key, ? super Signal<? extends Value>>` I think.
That is, it would allow:
-  maps with more generic keys
- maps with more generic value type, eg.`Object` instead of `Signal`
- **but** if map's value type is a `Signal` it must be one of elements compatible with `Value`

Reply:
Yeah, it tried to provide such things, but from the other side  what if we have the Map like: 
```
Map<Key, MyAwasomeSignal> cache = ...
```
and at some point user decided to do to something like 

```
MyAwasomeSignal sig = cache.get("someKey");
```

and potentially at this point, we may have `ClassCastException`

Reply:
@OlegDokuka wouldn't that be prevented from the fact that the V part of the `Map` is `? super Signal` -> the map contract doesn't guarantee that you get `MyAwesomeSignal`, just a `Signal` upper bound.

Reply:
Yup, you are right ... this generics always confuses my mind ...

Reply:
But, how we make sure, that cached value is a Signal? Actually, in that case we will be obligated to check instance type, or so(or am I wrong?). From the other side `dematerialize` may do it for us.

Reply:
Also, I like the idea of cache reader cache writer

Reply:
Actually, I will be really happy if we replace direct map with something like :
```java
Interface CacheReader<Key, Value> extends Function<Key, Mono<Value>>{}

Interface CacheWriter<Key,Value> extends BiFunction<Key,Value,Mono<Void>>{}
```

This give us enough flexibility to support adapter for in memory caches and distributet as well /cc @ben-mensen

Reply:
I don't think we'd need new interfaces (`Function` and `BiConsumer` would work IMO). Provide the simple `Map` facade + the advanced reader/writer method.

Also, requiring the user to deal with producing the `Mono` is the boilerplate we want to avoid, so I'd probably stick with `Signal<Value>`.

Reply:
Yeah, to avoid user boilerplate we may provide custom adapter, but the main reason to provide such interfaces in by supporting feature distributed caches

Reply:
> I will be really happy if we replace direct map with something like...

I agree that's better than a `Map`, because despite being really good it doesn't fit easily enough onto enough options. I could go either way on new types, though might start without them and let it evolve a bit to see if they'd be needed. If so after a release, adapting with an additional method shouldn't be too difficult.

Reply:
Okay, agree, totally make sense!

Reply:
I just meant the interfaces seemed a bit redundant where existing functional interfaces from Java 8 would seem to do the job just fine... That said I can see the benefit of having a `Mono` in the mix to adapt to/from something like caffeine's `AsyncLoadingCache`, and having alias interfaces can help readability too.

After toying a bit with the builder and reader/writer interfaces, here's a final API proposal:
```java
@FunctionalInterface
	interface MonoCacheReader<KEY, VALUE> extends Function<KEY, Mono<Signal<VALUE>>> {}

	@FunctionalInterface
	interface MonoCacheWriter<KEY, VALUE> extends BiFunction<KEY, Signal<VALUE>, Mono<Signal<VALUE>>> {}

	@FunctionalInterface
	interface MonoCacheBuilderCacheMiss<KEY, VALUE> {
		MonoCacheBuilderCacheWriter<KEY, VALUE> onCacheMissResume(Mono<VALUE> other);
	}

	@FunctionalInterface
	interface MonoCacheBuilderCacheWriter<KEY, VALUE> {
		Mono<VALUE> andWriteWith(MonoCacheWriter<KEY, VALUE> writer);
	}

	@FunctionalInterface
	interface MonoCacheBuilderMapMiss<VALUE> {
		Mono<VALUE> onCacheMissResume(Mono<VALUE> other);
	}

	public static <KEY, VALUE> MonoCacheBuilderMapMiss<VALUE> lookupMono(Map<KEY, Signal<VALUE>> cache, KEY key) {
		return other ->
				//TODO call lookupMono with adapters to Map methods
	}

	public static <KEY, VALUE> MonoCacheBuilderCacheMiss<KEY, VALUE> lookupMono(MonoCacheReader<KEY, VALUE> reader, KEY key) {
		return other -> writer ->
//TODO apply reader, dematerialize. if empty, materialize other then apply writer then dematerialize
	}
```

That would be used like this:
```java
	@Test
	public void test() {
		Mono<String> cached =
				lookupMono(s -> /* async loading Signal from cache */ Mono.just(Signal.next(s)), "foo")
						.onCacheMissResume(Mono.just("foo_value"))
						.andWriteWith((key, signalOfString) -> /* async storing in cache */ Mono.empty());

		Map<String, Signal<String>> map = Collections.singletonMap("foo", Signal.next("bar"));
		Mono<String> cachedInMap =
				lookupMono(map, "foo")
						.onCacheMissResume(Mono.just("foo_value"));
	}
```

Reply:
@OlegDokuka please use these names / signatures and implement the 2 lookup methods, and I'll merge this right away (I may try to improve the generic types later on)